### PR TITLE
Convert Form components to typescript

### PR DESCRIPTION
**Pull request by fzaninotto**
Refs #4505

### CODE REVIEW

File: packages/ra-ui-materialui/src/form/SimpleFormIterator.tsx [hunk 1]
```diff
@@ -85,7 +92,7 @@ const DefaultRemoveButton = props => {
     );
 };
 
-const SimpleFormIterator = props => {
+const SimpleFormIterator: FC<any> = props => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/SimpleFormIterator.tsx:95]: somehow I got lazy in the end

Comment (djhi) [reply packages/ra-ui-materialui/src/form/SimpleFormIterator.tsx:95]: :scream:

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 2]
```diff
@@ -150,17 +152,17 @@ const FormWithRedirect: FC<FormWithRedirectOwnProps & FormProps> = ({
 export interface FormWithRedirectOwnProps {
     defaultValue?: any;
     record?: Record;
-    save: (
+    save?: (
```
Comment (djhi) [inline packages/ra-core/src/form/FormWithRedirect.tsx:155]: We really should rename this to `onSubmit` imo

Comment (fzaninotto) [reply packages/ra-core/src/form/FormWithRedirect.tsx:155]: that'd be a breaking change... Perhaps in a future major.

Comment (djhi) [reply packages/ra-core/src/form/FormWithRedirect.tsx:155]: Support both with a warning ?

**Review by djhi — APPROVED** [review]

**Pull request merged**